File: ruby/easy/sqrt_x.rb
# @param {Integer} x
# @return {Integer}
def my_sqrt(x)
    return bs(1, x, x);
end

def bs(min, max, x)
    return min - 1 if min > max
    middle = (min + max)/2;
    sqr = middle * middle;
    if(sqr == x)
        return middle
    elsif(sqr > x)
        return bs(min, middle-1, x)
    else
        return bs(middle+1, max, x)
    end
end

File: javascript/easy/sqrtX.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    let start = 1, end = x;
    while(start <= end) {
        let middle = (start + end)/2;
        let sqr = middle * middle;
        if(sqr == x) {
            return middle;
        } else if(sqr > x) {
            end = middle - 1;
        } else {
            start = middle + 1;
        }
    }
    return start-1;
};
